### PR TITLE
More Validation

### LINKED ISSUE #87: Make use of some scim types in the library when it's possible
All of the scim types from the specs
https://tools.ietf.org/html/rfc7644#section-3.12
are pre-defined in the library but not all of them are used.

I believe that it's possible to cover some of them on the side of the library as it has a validation logic.

- invalidFilter
Currently `errors.ScimErrorBadParams` is returned but it should be just `errors.ScimErrorInvalidFilter`.

- mutability
Mutability is not checked. It should return `mutability` error in case when user tries to change immutable attributes that are defined immutable via the schema.

- invalidPath
Path isn't validated but might be validated with using the schema.

### PR DESCRIPTION
**Pull request by di-wu**
Closes #87 

- [x] Fix `invalidFilter`.
- [x] Check `immutable` attributes.
- [x] Check `invalidPath`.

### CODE REVIEW

**Comment by di-wu** [comment]
Filters can be complex, I want to analyse this and provide a better structure before adding checks.

File: README.md [hunk 1]
```diff
@@ -109,6 +109,43 @@ server := Server{
 log.Fatal(http.ListenAndServe(":8080", server))
 ```
 
+### Filtering
+In the `ListRequestParams` a parsed `Filter` can be found. This is just the filter query from the incoming request.
+This structure provides a two methods:
+
+- `Reduce`: removes all resources from a []map that do not pass the filter.
+- `IsValid`: checks whether the given map is passed the filter.
```
Comment (jdeflander) [inline README.md:117]: Is there a specific reason for designing the API like this? Would it also be possible to just filter the resources before they reach the handlers, and to not provide a parsed filter at all?

File: errors/error.go [hunk 2]
```diff
@@ -204,20 +204,29 @@ var (
...
+				Status: http.StatusInternalServerError,
+			}
 		}
-	}
-	if !checkApplicability(scimErr, method) {
-		return ScimError{
-			Detail: fmt.Sprintf("The HTTP status code %d is not applicable to the %s-operation.", scimErr.Status, method),
-			Status: http.StatusInternalServerError,
+		return scimErr
+	case *ScimError:
```
Comment (jdeflander) [inline errors/error.go:216]: Why do we want to support this case?

File: filter/comp.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
+package filter
```
Comment (jdeflander) [inline filter/comp.go:1]: You can panic instead of returning `invalid` for unreachable code.

File: filter/filter.go [hunk 4]
```diff
@@ -0,0 +1,166 @@
+package filter
```
Comment (jdeflander) [inline filter/filter.go:1]: Should this package be part of the public API?

File: README.md (hunk 1 above)
Comment (di-wu) [reply README.md:117]: I do not want to force people to use this implementation.
I think it is much more interesting to write a transpiler ([example](github.com/articulate/scim-filter-transpiler)).

Comment (di-wu) [reply README.md:117]: Filtering beforehand can also create conflicts w/ paging.

Comment (jdeflander) [reply README.md:117]: I see, but IMO we should try to keep the public API as small and simple as possible. If we don't want to implement actual filtering ourselves (which is indeed a good idea), we should just expose an AST of the parsed filter without the additional helper methods.

Comment (di-wu) [reply README.md:117]: I will keep it internal for now. If requested we can think about a better public API.

**Comment by di-wu** [comment]
> This PR adds a lot of exported methods to our public API, is that necessary?

The schema package provides a way to define compliant schemas and can be used to validate resources. The schema also needs to be accessible to validate filters and other paths. This was not possible because all the attribute fields were private.

File: README.md [hunk 5]
```diff
@@ -109,6 +109,43 @@ server := Server{
 log.Fatal(http.ListenAndServe(":8080", server))
 ```
 
+### Filtering
+In the `ListRequestParams` a parsed `Filter` can be found. This is just the filter query from the incoming request.
+This structure provides a two methods:
```
Comment (di-wu) [inline README.md:114]: This is not public anymore.

File: go.sum [hunk 6]
```diff
@@ -1,7 +1,9 @@
 github.com/davecgh/go-spew v1.1.0 h1:ZDRjVQ15GmhC3fiQ8ni8+OwkZQO4DARzQgrnXU1Liz8=
 github.com/davecgh/go-spew v1.1.0/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
-github.com/di-wu/scim-filter-parser v0.0.0-20191024153148-b33ea97ae796 h1:pzRK8hH/UoxQn2FqqUSYZwvKOvWMuNKyAurHVu2oU0M=
-github.com/di-wu/scim-filter-parser v0.0.0-20191024153148-b33ea97ae796/go.mod h1:Q9X8NLzgdqI3Z0XcbiPpdAEql664lC6ExBN7B6ZLhco=
+github.com/di-wu/scim-filter-parser v0.0.0-20200731082237-b6c96c1ee324 h1:Vf0SI9wRbvR+Y9jQmVF/c4UlH/6fQ7/09GbOfc2yMZ4=
+github.com/di-wu/scim-filter-parser v0.0.0-20200731082237-b6c96c1ee324/go.mod h1:jrab8v6Ax/lJkC4XLnRQW6RuH+UxXRDL3BrNDg6qdd0=
+github.com/di-wu/scim-filter-parser v0.0.0-20200731092905-86ddcb7e62bd h1:a5k7fCAnlAbzSFVV5j/aClknO+e08FnmW4zZcnfmHaE=
```
Comment (di-wu) [inline go.sum:5]: `go mod tidy`

File: handlers.go [hunk 7]
```diff
@@ -30,12 +33,26 @@ func (s Server) schemasHandler(w http.ResponseWriter, r *http.Request) {
...
-		resources = append(resources, v)
+		resource := v.ToMap()
+		if params.Filter != nil {
+			valid, err := filter.IsValid(resource)
+			if err != nil {
+				scimErr := errors.CheckScimError(err, http.MethodGet)
+				errorHandler(w, r, &scimErr)
+				return
+			}
+			fmt.Println(resource["id"])
```
Comment (di-wu) [inline handlers.go:50]: This does not need to be here...

**Pull request merged**